packages/contracts/contracts/maci/Tally.sol

```diff
@@ -0,0 +1,321 @@
+  ) public override isInitialized onlyOwner whenNotPaused afterCooldown {
+    uint256 amountLength = amounts.length;
+    uint256 totalFunds = totalAmount;
+    uint256 sum = 0;
+
+    for (uint256 index = 0; index < amountLength; ) {
+      uint256 amount = amounts[index];
+      sum += amount;
+
+      if (sum > totalFunds) {
+        revert InvalidWithdrawal();
+      }
+
+      totalAmount -= amount;
+
+      unchecked {
+        index++;
+      }
+    }
+
+    for (uint256 index = 0; index < amountLength; ) {
+      uint256 amount = amounts[index];
+      address receiver = receivers[index];
+
+      if (amount > 0) {
+        token.safeTransfer(receiver, amount);
+      }
+
+      unchecked {
+        index++;
+      }
+    }
+  }
+
+  /// @inheritdoc TallyBase
+  function addTallyResults(
+    uint256[] calldata voteOptionIndices,
+    uint256[] calldata results,
+    uint256[][][] calldata tallyResultProofs,
+    uint256 tallyResultSalt,
+    uint256 spentVoiceCreditsHashes,
+    uint256 perVOSpentVoiceCreditsHashes
+  ) public override isInitialized whenNotPaused afterVotingDeadline onlyOwner {
+    if (recipientCount == 0) {
+      recipientCount = registry.recipientCount();
+    }
+
+    if (recipientCount == totalTallyResults) {
+      revert TooManyResults();
+    }
+
+    super.addTallyResults(
+      voteOptionIndices,
+      results,
+      tallyResultProofs,
+      tallyResultSalt,
+      spentVoiceCreditsHashes,
+      perVOSpentVoiceCreditsHashes
+    );
+
+    if (recipientCount < totalTallyResults) {
+      revert TooManyResults();
+    }
+  }
+
+  /// @inheritdoc TallyBase
+  function addTallyResult(
+    uint256 voteOptionIndex,
+    uint256 tallyResult,
+    uint256[][] calldata tallyResultProof,
+    uint256 tallyResultSalt,
+    uint256 spentVoiceCreditsHash,
+    uint256 perVOSpentVoiceCreditsHash,
+    uint8 voteOptionTreeDepth
+  ) internal override {
+    super.addTallyResult(
+      voteOptionIndex,
+      tallyResult,
+      tallyResultProof,
+      tallyResultSalt,
+      spentVoiceCreditsHash,
+      perVOSpentVoiceCreditsHash,
+      voteOptionTreeDepth
+    );
+
+    totalVotesSquares += tallyResult ** 2;
+  }
+
+  /// @inheritdoc IPayoutStrategy
+  function claim(
+    IPayoutStrategy.Claim calldata params
+  ) public override isInitialized whenNotPaused afterVotingDeadline {
+    uint256 amount = getAllocatedAmount(params.voiceCreditsPerOption, params.totalSpent, params.tallyResult);
+    totalAmount -= amount;
+
+    bool isValid = verifyTallyResult(
+      params.index,
+      params.tallyResult,
+      params.tallyResultProof,
+      params.tallyResultSalt,
+      params.voteOptionTreeDepth,
+      params.spentVoiceCreditsHash,
+      params.perVOSpentVoiceCreditsHash
+    );
+
+    if (!isValid) {
+      revert InvalidTallyVotesProof();
+    }
+
+    IRecipientRegistry.Recipient memory recipient = registry.getRecipient(params.index);
+
+    token.safeTransfer(recipient.recipient, amount);
```

Comment:
seems like a recipient can keep calling this and drain all the money, we should be storing that they already claimed